spec/dynamic/layout.md

```diff
@@ -0,0 +1,335 @@
+|`fn()`[^3]| any[^1] | `<=ssize`[^1] | Non-null[^4] |
+| `Option<fn()>`[^3]| any[^1] | `<=size`[^1] | When initialized |
+| `[T;N]` | `size_of::<T>()*N` | Same as `T` | Each element valid |
+| `(T0,T1,..Tn)`[^6] | At least total of `T0`, `T1`, .. `Tn` + padding | At least maximum of `T0`, `T1`, ..`Tn` | Each element valid |
+| `()` | `0` | `1`  | Always |
+| `struct Foo{f0: T0, f1: T1, ..fN: Tn}`[^6][^7] | At least total of `T0`, `T1`, .. `Tn` + padding | At least maximum of `T0`, `T1`, ..`Tn` | Each field valid |
+| `struct Foo{}`[^7][^8] | `0` | `1` | Always |
+| `union Foo{f0: T0, f1: T1, ..fN; Tn}`[^6] | At least maximum of `T0`, `T1`, ..`Tn` | At least maximum of `T0`, `T1`, ..`Tn` |  |
+
+[^1]: Target defined
+
+[^2]: `usize` and pointers to `Thin` types have the same size and alignment requirement. Neither type has any padding bytes.
+
+[^3]: `fn()` refers to any fn-pointer type, regardless of return type, arguments, or ABI Tag
+
+[^4]: Range of valid values when computing as a suitably sized integer type (or pointer). Implicitly imposes a constraint of "When initialized"
+
+[^5]: References, and `Option`s of references (when `Some`) have additional aliasing constraints, see r#[dynamic.alias].
+
+[^6]: `repr(Rust)` types have an unspecified and unstable layout
+
+[^7]: Same as equivalent tuple struct. Struct size and alignment requirement may be modified by the `#[repr(align)]`
+
+[^8]: Same as equivalent unit struct
+
+## Layout Properties [§](r[dynamic.layout.properties])
+
+[§](r[dynamic.layout.properties.size]) Each `Sized` type has a property called it's size. The size is a non-negative integer.
+
+[!NOTE]: For example, the size of `u8` is `1`, and the size of `u32` is `4`.
+
+[§](r[dynamic.layout.properties.align]) Each `Sized` type and the slice type has a property called its alignment requirement.  The alignment requirement is a positive integer power of 2. An address satisfies an alignment requirement if it is a multiple of that alignment requirement. 
+
+[!NOTE]: The alignment requirement may simply be called the alignment or align of the type.
+
+[§](r[dynamic.layout.properties.size-align-value]) The size and alignment of any type is representible as a value of type `usize`. The alignment of a type divides its size.
+
+[!NOTE]: The minimum size of a type is `0`, and the minimum alignment requirement is `1`. A type that has size `0` and alignment requirement `1` is called a 1-ZST.
+
+[§](r[dynamic.layout.properties.storage]) Storage for a value is suitable for storing a value of a `Sized` type if it is at least the size of that type and the address of the start of the storage satisfies the alignment requirement.
+
+[§](r[dynamic.layout.properties.representation]) Each `Sized` type has a representation which determines how particular values of that type are layed out when read from or written to suitable storage. The representation of a type is exactly its size.
```

Comment:
> The representation of a type is exactly its size.

I don't know what this means, a representation is a lot more than just the size?

From this definition I can't even tell if you mean what t-opsem calls "layout" or what we call "representation". I think you need to be more clear about what the types are of the object that you are defining.

"representation", in the t-opsem meaning, is a *relation* between values and byte lists. It encodes a ton of things, not just layout aspects but also validity.

"layout" is much more first-order, it stores concrete data. For instance, the "layout" of a struct is given by the offsets of all its fields as well as the total size and alignment of the struct. I think this chapter should only do layout, not representation (and therefore also not validity). Something like:

Each `Sized` type has an associated *layout* that stores all the information required to lay out the type in memory.

Note: The layout is then used by the [representation] to determine how [values] are related to lists of bytes that represent the value in memory.


(I think unsized types may actually also have layout but I'm okay with deferring this to later, I haven't yet explored unsized types in MiniRust)

Reply:
This is trying to encode that the list of bytes that any value becomes is exactly as long as the size property of the type.

Reply:
This is trying to prescribe that the list of bytes that any value becomes is exactly as long as the size property of the type.